subsys/bluetooth/host/gatt.c

```diff
@@ -1211,73 +1300,129 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 	cfg->value = 0U;
 }
 
-#if defined(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE)
-static struct gatt_ccc_store {
-	struct bt_conn *conn_list[CONFIG_BT_MAX_CONN];
+#if defined(CONFIG_BT_SETTINGS_DELAYED_STORE)
+struct ds_peer {
+	uint8_t id;
+	bt_addr_le_t peer;
+
+	ATOMIC_DEFINE(flags, DELAYED_STORE_NUM_FLAGS);
+};
+
+static struct gatt_delayed_store {
+	struct ds_peer peer_list[CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN];
 	struct k_work_delayable work;
-} gatt_ccc_store;
+} gatt_delayed_store;
 
-static bool gatt_ccc_conn_is_queued(struct bt_conn *conn)
+struct ds_peer *gatt_delayed_store_find(uint8_t id, const bt_addr_le_t *peer_addr)
 {
-	return (conn == gatt_ccc_store.conn_list[bt_conn_index(conn)]);
+	struct ds_peer *el;
+
+	for (size_t i = 0; i < ARRAY_SIZE(gatt_delayed_store.peer_list); i++) {
+		el = &gatt_delayed_store.peer_list[i];
+		if (el->id == id &&
+		    bt_addr_le_eq(peer_addr, &el->peer)) {
+			return el;
+		}
+	}
+
+	return NULL;
 }
 
-static void gatt_ccc_conn_unqueue(struct bt_conn *conn)
+struct ds_peer *gatt_delayed_store_alloc(uint8_t id, const bt_addr_le_t *peer_addr)
 {
-	uint8_t index = bt_conn_index(conn);
+	struct ds_peer *el;
+
+	for (size_t i = 0; i < ARRAY_SIZE(gatt_delayed_store.peer_list); i++) {
+		el = &gatt_delayed_store.peer_list[i];
 
-	if (gatt_ccc_store.conn_list[index] != NULL) {
-		bt_conn_unref(gatt_ccc_store.conn_list[index]);
-		gatt_ccc_store.conn_list[index] = NULL;
+		/* Checking for the flags is cheaper than a memcmp for the
+		 * address, so we use that to signal that a given slot is
+		 * free.
+		 */
+		if (atomic_get(el->flags) == 0) {
+			bt_addr_le_copy(&el->peer, peer_addr);
+			el->id = id;
+
+			return el;
+		}
 	}
+
+	return NULL;
 }
 
-static void gatt_ccc_conn_enqueue(struct bt_conn *conn)
+static void gatt_delayed_store_free(struct ds_peer *el)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
-		/* Store the connection with the same index it has in
-		 * the conns array
-		 */
-		gatt_ccc_store.conn_list[bt_conn_index(conn)] =
-			bt_conn_ref(conn);
+	if (el) {
+		el->id = 0;
+		memset(&el->peer, 0, sizeof(el->peer));
+		atomic_set(el->flags, 0);
+	}
+}
 
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+static void gatt_delayed_store_enqueue(uint8_t id, const bt_addr_le_t *peer_addr,
+				       enum delayed_store_flags flag)
+{
+	bool bonded = bt_addr_le_is_bonded(id, peer_addr);
+	struct ds_peer *el = gatt_delayed_store_find(id, peer_addr);
+
+	if (bonded) {
+		if (el == NULL) {
+			el = gatt_delayed_store_alloc(id, peer_addr);
+			__ASSERT(el != NULL, "Can't save CF / CCC to flash");
+		}
+
+		atomic_set_bit(el->flags, flag);
+
+		k_work_reschedule(&gatt_delayed_store.work,
+				  K_MSEC(CONFIG_BT_SETTINGS_DELAYED_STORE_MS));
```

Comment:
Nothing will be stored to flash if the device reboots/loses power before `CONFIG_BT_SETTINGS_DELAYED_STORE_MS`?

Reply:
That's true. Proper fix would require to write from RX thread in some places I think (otherwise we never know if the value would actually be stored before response is sent). I noticed that implementation tries to avoid it.

Reply:
Yeah, not sure if there is a better way to solve it. It can anyway wait until laters as it was not introduced here as you say.